### PR TITLE
Examples: add Llama-3.1-8B

### PR DESCRIPTION
**Pull request by erwei-xilinx**
Stacked on #119 — review that first; this diff is only the 8B on top.

The third model on the shared harness, and the 1B's forward again: 32 layers of
4096, SwiGLU, one norm pair per block, no qk-norm. Nothing new in the prefill.
What it exercises is **size** — 32 resident decode layers, and a bf16
dequantization of an 8B on the host (~16 GB, which fits).

## Two per-model facts it does not share, both silent when wrong

**Its LM head is untied.** The 1B and 3B tie theirs to the embedding, and
`Q4nxModel.embed_norm_lmhead()` returns the full-precision embedding as the head
to match. The 8B bundle carries a real quantized `lm_head.weight`. Using the
embedding instead still yields a plausible logit vector and fluent text — it
cost a gate failure at 57618 to notice. mlir-air's loader says it outright:

```python
# UNTIED: a real quantized LM head, dequantized like any other projection.
lm_head = np.asarray(qm.dequant("lm_head.weight", config.vocab_size, config.emb_dim), bfloat16)
```

**Its gate prompt is different, deliberately.** "The capital of France is" ranks
" a" (264) fractionally above " Paris" (12366) on this model, and the HF bf16
reference agrees — so the siblings' prompt would fail a *correct* build.
mlir-air uses "The capital city of France is called", where " Paris" leads by
3.4 logits. Copied with the reason, because sharing one prompt across the family
is the obvious simplification and the trap.

## A third constant that had been matching by luck

`DECODE_STACK=8064` lowers the AIE core stack from the builder's 10240 default.
`decode_build` now reads `fused_decode.STACK_SIZE` after importing the builder
under the model's environment, exactly as mlir-air's own driver does
(`XRTBackend(stack_size=STACK_SIZE)`), instead of taking `FusedDecodeOp`'s
default. The 1B and 3B are unaffected — they take that same default — so this is
the first model where it differs.

That makes three per-model constants in this family that were matching by
coincidence: `-DMODEL_TYPE` (#118), `GLU_SLICE_EXPECTED` (#118, still 1024 for
all three), and now the stack. All three are derived or recorded rather than
assumed.

## Verification

On Strix (aie2p), against a clean sparse checkout at the pinned commit:

| | |
|---|---|
| `decode_L2048.insts.bin` vs mlir-air's own 8B driver | byte-identical |
| its xclbin | 68 bytes differ — all UUID / UniqueID / TimeStamp |
| stack actually used | 8064, from the builder |
| prefill gate | 12366 |
| 30 greedy tokens | coherent, 9.5 tok/s (mlir-air's README: 10.7) |
| 1B and 3B gates | unaffected |

Output: `' Paris. It is located in the northern part of the country, in the region
of Île-de-France. Paris is a major city in Europe'`

## On the generated ids

Worth stating precisely, because the headline number is not 30/30.

Against mlir-air's own driver, same prompt, 30 greedy tokens:

| route | vs mlir-air |
|---|---|
| our Triton prefill -> their decode | **27/30**, diverging at token 27 |
| their decode-replay, through our artifacts | **30/30** |

The second row is the one that tests this PR: with both sides building the
prompt KV the same way, our decode artifacts and loading reproduce mlir-air's
output exactly, for every token.

The first row differs because the two runs build the prompt KV by different
routes — ours a batched Triton prefill, theirs a token-by-token decode replay —
which are numerically distinct by construction (bf16 rounding, different
accumulation order). After 27 identical tokens a near-tie flips. Our own run is
deterministic across repeats, and the first-token gate, which is the actual
correctness signal, is exact.

For contrast the 3B agreed 40/40 across those same two routes; the 8B is deeper
and tips sooner. Nothing here suggests a defect, but it is not the clean
identity the 3B PR could claim, so I am not claiming it.

### CODE REVIEW

File: examples/llama31_8b_q4nx/Makefile [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+# prefill's kernels as PDI, and the decode with its instruction stream patched
+# per token (../llm_q4nx/hsa_decode.py). Correct, and currently ~22 tok/s against XRT's
+RUNTIME  ?= xrt
+export AMD_TRITON_NPU_RUNTIME = $(RUNTIME)
+DECODE_FORMAT := $(if $(filter hsa,$(RUNTIME)),pdi,xclbin)
+
+OPS      ?= all
+BACKEND  ?= npu
+N_TOKENS ?= 200
+PROMPT   ?= The capital of France is
```
[the remark this answers is not in the document]
Comment (erwei-xilinx) [reply examples/llama31_8b_q4nx/Makefile]: Fixed in cc71945, and this mattered more than it looks. The harness only applies the gate when the ids match `config.PROMPT`, so with the siblings' prompt as the default, `make run` was silently running **ungated** on this model. The default is now `The capital city of France is called`, verified to tokenize to exactly `config.PROMPT` ([128000, 791, 6864, 3363, 315, 9822, 374, 2663]) through the same tokenizer resolution the harness uses.

While in this file I also removed `make chat` and corrected the `RUNTIME=hsa` comment: this model's driver names its decoder `FusedDecode8B`, so the harness refuses both up front and each target could only ever exit 1. Same defect just removed from the 3B.

File: examples/llama31_8b_q4nx/config.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    headers, which do not encode each projection's logical [out, K], so the
+    dims have to be supplied. `_proj_dims` derives them from the config, and is
+    imported rather than rewritten -- getting it wrong transposes a projection,
+    which loads and then produces fluent nonsense.
+
+    Returns a dict with
+        layers: list of 28 dicts, each with
+            attn_norm, ffn_norm : float32 [D]
+            q, k, v, o, gate, up, down : bfloat16 [K, out]  (y = x @ W)
+        embed, final_norm, lm_head : float32
```
[the remark this answers is not in the document]
Comment (erwei-xilinx) [reply examples/llama31_8b_q4nx/config.py:99]: Fixed in cc71945 -- 32, not 28. Inherited when this example was derived from the 3B.

File: examples/llama31_8b_q4nx/config.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+def rope_lut(seq_len, dtype=None):
+    """[seq_len, DH] = [cos_0..cos_63, sin_0..sin_63] per position.
+
+    mlir-air's generate_rope_lut, unmodified. It carries the llama3 frequency
+    scaling (factor 32, old_ctx 8192) that the fused decode also applies; a
+    reimplementation that drops it produces K rotated on the wrong frequencies
+    and long prompts silently degrade.
+
+    The 3B's table is twice the 1B's width, because its heads are: 128 wide, so
+    64 cos and 64 sin.
```
[the remark this answers is not in the document]
Comment (erwei-xilinx) [reply examples/llama31_8b_q4nx/config.py]: Fixed in cc71945. Rewritten to describe the table by its own head width rather than by a sibling: 64 cos and 64 sin for this model's 128-wide heads, the same shape as the 3B's and twice the 1B's.

File: examples/llama31_8b_q4nx/llama31_8b_q4nx_inference.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+The prefill is this repo's -- Triton kernels on XDNA (prefill.py, and
+../llm_q4nx/kernels.py). The decode is mlir-air's fused Q4NX decode, run
+unmodified: the shared harness writes the KV handoff npz that its `generate()`
+already loads, and neutralizes only the step that would have produced it.
+
+Llama-3.1-8B is architecturally the 1B -- SwiGLU, one norm pair per block, no
+qk-norm -- so it shares the 1B's forward as well as the harness. Everything
+3B-specific is in `config.py`: 28 layers of 3072, 128-wide heads, and its own
+weight bundle.
```
[the remark this answers is not in the document]
Comment (erwei-xilinx) [reply examples/llama31_8b_q4nx/llama31_8b_q4nx_inference.py]: Fixed in cc71945 -- 32 layers of 4096. The sentence now also names the two things that actually distinguish this model from its siblings, the untied LM head and its own gate prompt, since that is what a reader of this file needs.

File: examples/llm_q4nx/llama_prefill.py [hunk 5]
```diff
@@ -282,7 +292,12 @@ def load_weights(self, model=None):
         self.final_norm = _t(raw["final_norm"])
         self.lm_head = _t(raw["lm_head"], torch.bfloat16)  # tied to embed
```
[the remark this answers is not in the document]
Comment (erwei-xilinx) [reply examples/llm_q4nx/llama_prefill.py]: Fixed in cc71945, and this is the sharpest of the six. That comment was true when only the 1B and 3B existed and this PR makes it false -- it is precisely the assumption that cost a debugging session here: the 8B ran on the tied assumption and generated 57618 instead of " Paris". The comment now says which models tie, which does not, and that `config.load_q4nx` is what decides.

File: examples/llm_q4nx/llama_prefill.py [hunk 6]
```diff
@@ -282,7 +292,12 @@ def load_weights(self, model=None):
...
         self._w = []
-        for L in raw["layers"]:
+        # Pop as we go: `raw` holds every layer's numpy arrays at once, and
+        # holding those alongside the torch copies doubles the resident set for
+        # the length of the loop. Each layer's source is dead once converted.
+        layers = raw["layers"]
+        while layers:
+            L = layers.pop(0)
```
[the remark this answers is not in the document]
Comment (erwei-xilinx) [reply examples/llm_q4nx/llama_prefill.py]: Fixed in cc71945. Popping the tail alone would have reversed the layers, and that order is load-bearing -- layer k must land at `_w[k]` -- so it reverses once and then pops: O(1) per iteration, original order preserved. Asserted by test rather than by eye, and the 1B and 8B prefill gates were re-run (12366 both).
